src/Analyzers/CSharp/Analyzers/UseUTF8StringLiteral/UseUTF8StringLiteralDiagnosticAnalyzer.cs

```diff
@@ -0,0 +1,150 @@
+                EnforceOnBuildValues.UseUTF8StringLiteral,
+                CSharpCodeStyleOptions.PreferUTF8StringLiteral,
+                LanguageNames.CSharp,
+                new LocalizableResourceString(nameof(CSharpAnalyzersResources.Convert_to_UTF8_string_literal), CSharpAnalyzersResources.ResourceManager, typeof(CSharpAnalyzersResources)),
+                new LocalizableResourceString(nameof(CSharpAnalyzersResources.Use_UTF8_string_literal), CSharpAnalyzersResources.ResourceManager, typeof(CSharpAnalyzersResources)))
+        {
+        }
+
+        protected override void InitializeWorker(AnalysisContext context)
+            => context.RegisterCompilationStartAction(context =>
+            {
+                if (context.Compilation.LanguageVersion() < LanguageVersion.Preview)
+                    return;
+
+                var expressionType = context.Compilation.GetTypeByMetadataName(typeof(System.Linq.Expressions.Expression<>).FullName!);
+
+                context.RegisterOperationAction(c => AnalyzeOperation(c, expressionType), OperationKind.ArrayCreation);
+            });
+
+        private void AnalyzeOperation(OperationAnalysisContext context, INamedTypeSymbol? expressionType)
+        {
+            var arrayCreationExpression = (IArrayCreationOperation)context.Operation;
+
+            // Don't offer if the user doesn't want it
+            var option = context.GetOption(CSharpCodeStyleOptions.PreferUTF8StringLiteral);
+            if (!option.Value)
+                return;
+
+            // Only replace arrays with initializers
+            if (arrayCreationExpression.Initializer is null)
+                return;
+
+            var byteType = context.Compilation.GetSpecialType(SpecialType.System_Byte);
+            var elementType = (arrayCreationExpression.Type as IArrayTypeSymbol)?.ElementType;
+            if (!SymbolEqualityComparer.Default.Equals(elementType, byteType))
+                return;
+
+            // UTF8 strings are not valid to use in attributes
+            if (arrayCreationExpression.Syntax.Ancestors().OfType<AttributeSyntax>().Any())
+                return;
+
+            // Can't use a UTF8 string inside an expression tree.
+            var semanticModel = context.Operation.SemanticModel;
+            Contract.ThrowIfNull(semanticModel);
+            if (arrayCreationExpression.Syntax.IsInExpressionTree(semanticModel, expressionType, context.CancellationToken))
+                return;
+
+            var values = arrayCreationExpression.Initializer.ElementValues.SelectAsArray(v => v.ConstantValue.HasValue && v.ConstantValue.Value is not null, v => v.ConstantValue.Value);
+
+            // If we couldn't get constant values for all elements then we can't offer
+            if (values.Length != arrayCreationExpression.Initializer.ElementValues.Length)
+                return;
+
+            var stringValue = GetStringValue(values);
```

Comment:
i think if you do a few more checks up front, you can avoid this in teh analzyer.  i think that would be preferable as i can imagine some very large embedded byte[]s that will force large allocations here.  moving that to hte fixer would be nice.

Reply:
Yeah, I didn't like how this has to create a byte array to get a ROS to pass to Rune. With that in place though, it felt like a pretty small jump to just keep things simple and produce the string, but you're probably right.

It would be great if the bytes could all be streamed direct from the operation nodes to Rune. Never really used ROS before though so need to dig in moreand work out how. Actually, could probably just create smaller 4 byte arrays, as it won't consume more than that.